pkg/cmd/admin/migrate/migrator.go

```diff
@@ -0,0 +1,414 @@
+package migrate
+
+import (
+	"fmt"
+	"io"
+	"strings"
+
+	"github.com/golang/glog"
+	"github.com/spf13/cobra"
+
+	"k8s.io/kubernetes/pkg/api/errors"
+	"k8s.io/kubernetes/pkg/api/meta"
+	"k8s.io/kubernetes/pkg/kubectl"
+	"k8s.io/kubernetes/pkg/kubectl/resource"
+	"k8s.io/kubernetes/pkg/util/sets"
+
+	cmdutil "github.com/openshift/origin/pkg/cmd/util"
+	"github.com/openshift/origin/pkg/cmd/util/clientcmd"
+)
+
+// MigrateVisitFunc is invoked for each returned object, and may return a
+// Reporter that can contain info to be used by save.
+type MigrateVisitFunc func(info *resource.Info) (Reporter, error)
+
+// MigrateActionFunc is expected to persist the altered info.Object. The
+// Reporter returned from Visit is passed to this function and may be used
+// to carry additional information about what to save on an object.
+type MigrateActionFunc func(info *resource.Info, reporter Reporter) error
+
+// MigrateFilterFunc can return false to skip an item, or an error.
+type MigrateFilterFunc func(info *resource.Info) (bool, error)
+
+// Reporter indicates whether a resource requires migration.
+type Reporter interface {
+	// Changed returns true if the resource requires migration.
+	Changed() bool
+}
+
+// ResourceOptions assists in performing migrations on any object that
+// can be retrieved via the API.
+type ResourceOptions struct {
+	In          io.Reader
+	Out, ErrOut io.Writer
+
+	Include       []string
+	Filenames     []string
+	AllNamespaces bool
+	Confirm       bool
+	Output        string
+	FromKey       string
+	ToKey         string
+
+	DefaultExcludes []string
```

Comment:
> Equivalent to. What would group resource add?

It's the type we actually use in all the internal code.  I'd rather have the type in `Options` be correct and transform the user input inside of `Complete`.  That way any late composition operates against correct and unambiguous types.